k2/python/k2/rnnt_loss.py

```diff
@@ -0,0 +1,995 @@
+      boundary:
+        a LongTensor of shape [B, 4] with elements interpreted as
+        [begin_symbol, begin_frame, end_symbol, end_frame] that is treated as
+        [0, 0, S, T]
+        if boundary is not supplied.
+        Most likely you will want begin_symbol and begin_frame to be zero.
+    Returns:
+        (px, py) (the names are quite arbitrary).
+           px: logprobs, of shape [B][S][T+1]
+           py: logprobs, of shape [B][S+1][T]
+        in the recursion:
+          p[b,0,0] = 0.0
+          p[b,s,t] = log_add(p[b,s-1,t] + px[b,s-1,t],
+                             p[b,s,t-1] + py[b,s,t-1])
+          .. where p[b][s][t] is the "joint score" of the pair of subsequences
+          of length s and t respectively.  px[b][s][t] represents the
+          probability of extending the subsequences of length (s,t) by one in
+          the s direction, given the particular symbol, and py[b][s][t]
+          represents the probability of extending the subsequences of length
+          (s,t) by one in the t direction,
+          i.e. of emitting the termination/next-frame symbol.
+
+          px[:,:,T] equals -infinity, meaning on the "one-past-the-last" frame
+          we cannot emit any symbols.  This is simply a way of incorporating
+          the probability of the termination symbol on the last frame.
+    """
+    assert (
+        lm.ndim == 3
+        and am.ndim == 3
+        and lm.shape[0] == am.shape[0]
+        and lm.shape[2] == am.shape[2]
+    )
+    (B, T, C) = am.shape
+    S = lm.shape[1] - 1
+    assert symbols.shape == (B, S)
+
+    # subtracting am_max and lm_max is to ensure the probs are in a good range
+    # to do exp() without causing underflow or overflow.
+    am_max, _ = torch.max(am, dim=2, keepdim=True)  # am_max: [B][T][1]
+    lm_max, _ = torch.max(lm, dim=2, keepdim=True)  # lm_max: [B][S+1][1]
+    am_probs = (am - am_max).exp()
+    lm_probs = (lm - lm_max).exp()
+    # normalizers: [B][S+1][T]
+    normalizers = (
+        torch.matmul(lm_probs, am_probs.transpose(1, 2)) + 1.0e-37
+    ).log()
+
+    # add lm_max and am_max to normalizers, to make it as if we had not
+    # subtracted am_max and lm_max above.
+    normalizers = normalizers + lm_max + am_max.transpose(1, 2)  # [B][S+1][T]
+
+    # px is the probs of the actual symbols..
+    px_am = torch.gather(
+        am.unsqueeze(1).expand(B, S, T, C),
+        dim=3,
+        index=symbols.reshape(B, S, 1, 1).expand(B, S, T, 1),
+    ).squeeze(
+        -1
+    )  # [B][S][T]
+
+    px_am = torch.cat(
+        (
+            px_am,
+            torch.full(
+                (B, S, 1), float("-inf"), device=px_am.device, dtype=px_am.dtype
+            ),
+        ),
+        dim=2,
+    )  # now: [B][S][T+1], index [:,:,T] has -inf..
+
+    if boundary is not None:
+        assert boundary.shape == (B, 4)
+        mask = torch.stack([torch.arange(0, T + 1, device=px_am.device)] * B)
```

Comment:
Here, instead of the torch.stack expression, it would be more efficient to do
`torch.arange(0, T + 1, device=px_am.device).reshape(1, T + 1).expand(B, T + 1)`
which is a zero-copy expression.
Also, below, instead of doing `mask = torch.where(mask < boundary[:, 3].reshape(B, 1), True, False)`
you can just use `mask = mask < boundary[:, 3].reshape(B, 1)` , which is already a boolean expression.